src/Storages/StorageDelta.cpp

```diff
@@ -0,0 +1,354 @@
+#include <Common/config.h>
+
+#if USE_AWS_S3
+
+#include <Storages/StorageDelta.h>
+#include <Common/logger_useful.h>
+
+#include <IO/ReadBufferFromS3.h>
+#include <IO/ReadSettings.h>
+#include <IO/S3Common.h>
+
+#include <Storages/StorageFactory.h>
+#include <Storages/checkAndGetLiteralArgument.h>
+
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+
+#include <QueryPipeline/Pipe.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int S3_ERROR;
+    extern const int BAD_ARGUMENTS;
+}
+
+void DeltaLakeMetadata::add(const String & key, uint64_t timestamp)
+{
+    file_update_time[key] = timestamp;
+}
+
+void DeltaLakeMetadata::remove(const String & key, uint64_t /*timestamp */)
+{
+    file_update_time.erase(key);
+}
+
+std::vector<String> DeltaLakeMetadata::ListCurrentFiles() &&
+{
+    std::vector<String> keys;
+    keys.reserve(file_update_time.size());
+
+    for (auto && [k, _] : file_update_time)
+    {
+        keys.push_back(k);
+    }
+
+    return keys;
+}
+
+JsonMetadataGetter::JsonMetadataGetter(StorageS3::S3Configuration & configuration_, const String & table_path_, Poco::Logger * log_)
+    : base_configuration(configuration_), table_path(table_path_), metadata(), log(log_)
+{
+    Init();
+}
+
+void JsonMetadataGetter::Init()
+{
+    auto keys = getJsonLogFiles();
+
+    // read data from every json log file
+    for (const String & key : keys)
+    {
+        auto buf = createS3ReadBuffer(key);
+        String json_str;
+        size_t opening(0), closing(0);
+        char c;
+
+        while (buf->read(c))
+        {
+            // skip all space characters for JSON to parse correctly
+            if (isspace(c))
+            {
+                continue;
+            }
+
+            json_str.push_back(c);
+
+            if (c == '{')
```

Comment:
Let's not parse it manually, see example how it can be done better here
https://github.com/ClickHouse/ClickHouse/blob/712aebc077291785f100cdacb65df98402bfc9cb/src/Storages/MeiliSearch/SourceMeiliSearch.cpp#L201-L206

Reply:
Changed parsing method: added methods to read one json from metadata file and method to process json